src/vmm/src/builder.rs

```diff
@@ -471,37 +470,60 @@ pub fn build_microvm_from_snapshot(
         vcpu_count,
     )?;
 
-    #[cfg(target_arch = "x86_64")]
-    // Check if we need to scale the TSC.
-    // We start by checking if the CPU model in the snapshot is
-    // the same as this host's. If they are the same, we don't
-    // need to do anything else.
-    if !is_same_model(&microvm_state.vcpu_states[0].cpuid) {
-        // Extract the TSC freq from the state.
-        // No TSC freq in snapshot means we have to fail-fast.
-        let state_tsc = microvm_state.vcpu_states[0]
-            .tsc_khz
-            .ok_or_else(|| {
-                MicrovmStateError::IncompatibleState(
-                    "Error configuring the TSC, frequency not present in snapshot.".to_string(),
-                )
-            })
-            .map_err(RestoreMicrovmState)?;
-
-        // Scale the TSC frequency for all VCPUs, if needed.
-        if vcpus[0]
-            .kvm_vcpu
-            .is_tsc_scaling_required(state_tsc)
-            .map_err(|err| MicrovmStateError::IncompatibleState(err.to_string()))
-            .map_err(RestoreMicrovmState)?
-        {
-            for vcpu in &vcpus {
-                vcpu.kvm_vcpu
-                    .set_tsc_khz(state_tsc)
-                    .map_err(|err| MicrovmStateError::IncompatibleState(err.to_string()))
-                    .map_err(RestoreMicrovmState)?;
+    // CPUID is supported when:
+    // - We are on an x86 archtecture with `see` enabled and `sgx disabled`.
+    // - We are on an x86_64 architecture with `sgx` disabled
+    #[cfg(any(
```

Comment:
We should do a runtime check of KVM_CAP_EXT_CPUID rather than this static check. Even if we do add this check, I believe we should do it at runtime since Firecracker can run on multiple machine types.

Reply:
I think this has been resolved?

When getting the KVM CPUID, I use:
```rust
kvm_ioctls::Kvm::new()?.get_supported_cpuid(kvm_bindings::KVM_MAX_CPUID_ENTRIES)?;
```
which I believe handles the case where the KVM does not support CPUID (looking at the underlying implementation it seems to do this, although not very well).

---
If needed, we could implement a check like:
```rust
/// Returns if `KVM_CAP_EXT_CPUID` is supported.
pub fn kvm_cap_ext_cpuid() -> std::result::Result<bool, vmm_sys_util::errno::Error> {
    let kvm = kvm_ioctls::Kvm::new()?;
    Ok(kvm.check_extension(kvm_ioctls::Cap::ExtCpuid))
}
```